### ISSUE TITLE
Odd inconsistency between GetSymbolInfo of invocation and its argument

### ISSUE THREAD

**Issue by Neme12** [dotnet/roslyn#30219]
```c#
using System;
using System.Collections.Concurrent;

class JToken { }

class C
{
    private static readonly ConcurrentDictionary<Type, Func<JToken, object>> _deserializeHelpers =
        new ConcurrentDictionary<Type, Func<JToken, object>>();

    private static void DeserializeObject(JToken token, Type type)
    {
        _deserializeHelpers.GetOrAdd(type, key => CreateDeserializeDelegate(key));
    }
}
```

There are 2 overloads of `GetOrAdd`:
1.  where the second parameter is of type `Func<JToken, object>`
![image](https://user-images.githubusercontent.com/11444821/46230158-01dbd780-c368-11e8-8f7a-31efd7bba346.png)
2. where the second parameter is of type `Func<Type, Func<JToken, object>>`
![image](https://user-images.githubusercontent.com/11444821/46230189-1ddf7900-c368-11e8-9403-093a8c1facaa.png)

`GetDeclaredSymbol` on the `key` lambda parameter returns a parameter whose type is `JToken` (suggesting this information comes the *first* overload).
`GetSymbolInfo` on the `GetOrAdd` invocation returns 2 candidate symbols. The first candidate symbol is the *second* overload. Is it an unreasonable assumption that the first candidate symbol would be the one used to gather parameter types? Should they not match? Or is there no guaranteed ordering of `CandidateSymbols`?

(this seems to be the root cause of #5338)

**Comment by Neme12** [comment]
@gafter @jcouv Is this the expected behavior?

**Comment by jcouv** [comment]
@Neme12 Thanks for the detailed issue report. I tried transferring it to a unittest, but I didn't quite manage to repro. Specifically, `GetSymbolInfo` on the lambda parameter gave me no type.
I tried a simpler repro, and then also one with `ConcurrentDictionary`, but behavior was the same.

```C#
        [Fact, WorkItem(30219, "https://github.com/dotnet/roslyn/issues/30219")]
        public void Test30219()
        {
            string source = @"
public class C<TKey, TValue>
{
    void M(TKey key, TValue value) => throw null;
    void M(TKey key, System.Func<TKey, TValue> delayedValue) => throw null;

    static void Method(C<int, System.Func<string, object>> c, int key)
    {
        c.M(key, k => Missing(k));
    }
}
";

            var comp = CreateCompilation(source);
            comp.VerifyDiagnostics(
                // (9,23): error CS0103: The name 'Missing' does not exist in the current context
                //         c.M(key, k => Missing(k));
                Diagnostic(ErrorCode.ERR_NameNotInContext, "Missing").WithArguments("Missing").WithLocation(9, 23)
                );
            var tree = comp.SyntaxTrees.Single();
            var model = comp.GetSemanticModel(tree, ignoreAccessibility: false);

            var lambdaParameter = tree.GetRoot().DescendantNodes().OfType<SimpleLambdaExpressionSyntax>().Single().Parameter;
            Assert.Equal("k", lambdaParameter.ToString());
            var lambdaParameterInfo = model.GetSymbolInfo(lambdaParameter);
            Assert.Equal(CandidateReason.None, lambdaParameterInfo.CandidateReason);

            var invocation = tree.GetRoot().DescendantNodes().OfType<InvocationExpressionSyntax>().First();
            Assert.Equal("c.M(key, k => Missing(k))", invocation.ToString());
            var invocationInfo = model.GetSymbolInfo(invocation);
            Assert.Equal(CandidateReason.OverloadResolutionFailure, invocationInfo.CandidateReason);
            Assert.Null(invocationInfo.Symbol);
            Assert.Equal("void C<System.Int32, System.Func<System.String, System.Object>>.M(System.Int32 key, System.Func<System.String, System.Object> value)",
                invocationInfo.CandidateSymbols[0].ToTestDisplayString());
            Assert.Equal("void C<System.Int32, System.Func<System.String, System.Object>>.M(System.Int32 key, System.Func<System.Int32, System.Func<System.String, System.Object>> delayedValue)",
                invocationInfo.CandidateSymbols[1].ToTestDisplayString());
        }

        [Fact, WorkItem(30219, "https://github.com/dotnet/roslyn/issues/30219")]
        public void Test30219_ConcurrentDictionary()
        {
            string source = @"
public class C<TKey, TValue>
{
    static void Method(System.Collections.Concurrent.ConcurrentDictionary<int, System.Func<string, object>> c, int key)
    {
        c.GetOrAdd(key, k => Missing(k));
    }
}
";

            var comp = CreateCompilation(source);
            comp.VerifyDiagnostics(
                // (6,30): error CS0103: The name 'Missing' does not exist in the current context
                //         c.GetOrAdd(key, k => Missing(k));
                Diagnostic(ErrorCode.ERR_NameNotInContext, "Missing").WithArguments("Missing").WithLocation(6, 30)
                );
            var tree = comp.SyntaxTrees.Single();
            var model = comp.GetSemanticModel(tree, ignoreAccessibility: false);

            var lambdaParameter = tree.GetRoot().DescendantNodes().OfType<SimpleLambdaExpressionSyntax>().Single().Parameter;
            Assert.Equal("k", lambdaParameter.ToString());
            var lambdaParameterInfo = model.GetSymbolInfo(lambdaParameter);
            Assert.Equal(CandidateReason.None, lambdaParameterInfo.CandidateReason);

            var invocation = tree.GetRoot().DescendantNodes().OfType<InvocationExpressionSyntax>().First();
            Assert.Equal("c.GetOrAdd(key, k => Missing(k))", invocation.ToString());
            var invocationInfo = model.GetSymbolInfo(invocation);
            Assert.Equal(CandidateReason.OverloadResolutionFailure, invocationInfo.CandidateReason);
            Assert.Null(invocationInfo.Symbol);
            Assert.Equal("System.Func<System.String, System.Object> System.Collections.Concurrent.ConcurrentDictionary<System.Int32, System.Func<System.String, System.Object>>.GetOrAdd(System.Int32 key, System.Func<System.Int32, System.Func<System.String, System.Object>> valueFactory)",
                invocationInfo.CandidateSymbols[0].ToTestDisplayString());
            Assert.Equal("System.Func<System.String, System.Object> System.Collections.Concurrent.ConcurrentDictionary<System.Int32, System.Func<System.String, System.Object>>.GetOrAdd(System.Int32 key, System.Func<System.String, System.Object> value)",
                invocationInfo.CandidateSymbols[1].ToTestDisplayString());
        }
```

**Comment by Neme12** [comment]
@jcouv I'm sorry, it was actually `GetDeclaredInfo` on the parameter, not `GetSymbolInfo` (I only checked this with the syntax visualizer which doesn't distinguish these and forgot that this is a declaration 😕 ). Thanks for trying to investigate this.

**Comment by jcouv** [comment]
No worries. Thanks for the clarification.

Below is the repro-as-unittest. I agree it does seem odd. 

Digging into this, I found that the selection of an overload for binding of error recovery case has little to do with the order of candidates from overload resolution. That selection is done by the method `GuessBestBoundLambda` (copied below, called from `CreateBadCall`) which selects the overload with `int` parameter. 

That ranking/guess currently has no effect on the order of `.CandidateSymbols` (built from `BoundCall.OriginalMethodsOpt` in `CreateReducedExtensionMethodsFromOriginalsIfNecessary`).

I'll check with Neal, but one approach could be to fix `CreateBadCall` to place the best guess at the top of the list in `BoundCall.OriginalMethodsOpt`.

```C#
        private static BoundLambda GuessBestBoundLambda<T>(ImmutableDictionary<T, BoundLambda> candidates)
        {
            switch (candidates.Count)
            {
                case 0:
                    return null;
                case 1:
                    return candidates.First().Value;
                default:
                    // Prefer candidates with fewer diagnostics.
                    IEnumerable<KeyValuePair<T, BoundLambda>> minDiagnosticsGroup = candidates.GroupBy(lambda => lambda.Value.Diagnostics.Length).OrderBy(group => group.Key).First();

                    // If multiple candidates have the same number of diagnostics, order them by delegate type name.
                    // It's not great, but it should be stable.
                    return minDiagnosticsGroup
                        .OrderBy(lambda => GetLambdaSortString(lambda.Value.Symbol))
                        .FirstOrDefault()
                        .Value;
            }
        }
```

```C#
        [Fact, WorkItem(30219, "https://github.com/dotnet/roslyn/issues/30219")]
        public void Test30219()
        {
            string source = @"
public class C<TKey, TValue>
{
    void M(TKey key, TValue value) => throw null;
    void M(TKey key, System.Func<TKey, TValue> delayedValue) => throw null;

    static void Method(C<int, System.Func<string, object>> c, int key)
    {
        c.M(key, k => Missing(k));
    }
}
";

            var comp = CreateCompilation(source);
            comp.VerifyDiagnostics(
                // (9,23): error CS0103: The name 'Missing' does not exist in the current context
                //         c.M(key, k => Missing(k));
                Diagnostic(ErrorCode.ERR_NameNotInContext, "Missing").WithArguments("Missing").WithLocation(9, 23)
                );
            var tree = comp.SyntaxTrees.Single();
            var model = comp.GetSemanticModel(tree, ignoreAccessibility: false);

            var lambdaParameter = tree.GetRoot().DescendantNodes().OfType<SimpleLambdaExpressionSyntax>().Single().Parameter;
            Assert.Equal("k", lambdaParameter.ToString());
            var lambdaParameterSymbol = model.GetDeclaredSymbol(lambdaParameter);
            Assert.Equal("System.Int32 k", lambdaParameterSymbol.ToTestDisplayString());

            var invocation = tree.GetRoot().DescendantNodes().OfType<InvocationExpressionSyntax>().First();
            Assert.Equal("c.M(key, k => Missing(k))", invocation.ToString());
            var invocationInfo = model.GetSymbolInfo(invocation);
            Assert.Equal(CandidateReason.OverloadResolutionFailure, invocationInfo.CandidateReason);
            Assert.Null(invocationInfo.Symbol);
            Assert.Equal("void C<System.Int32, System.Func<System.String, System.Object>>.M(System.Int32 key, System.Func<System.String, System.Object> value)",
                invocationInfo.CandidateSymbols[0].ToTestDisplayString());
            Assert.Equal("void C<System.Int32, System.Func<System.String, System.Object>>.M(System.Int32 key, System.Func<System.Int32, System.Func<System.String, System.Object>> delayedValue)",
                invocationInfo.CandidateSymbols[1].ToTestDisplayString());
        }
```

**Comment by jcouv** [comment]
Hum, there's a problem with the approach I was thinking about (tweaking `CreateBadCall` to sort candidates with guess from lambda error recovery).
What if we have an invocation with two bad lambdas?
Both lambdas/arguments will be sorted independently (inside `BuildArgumentsForErrorRecovery`), and it's possible their top choice won't agree. Then it's unclear what should be the top choice for the invocation... :-(

**Comment by Neme12** [comment]
Hm, I see. I didn't know that's the way the compiler works. It seems a bit odd to me that the binding of individual arguments could actually come from different overloads. I would expect there would be a single overload picked (for example to have the minimum sum of diagnostics in all lambdas in the invocation) and then every argument would have to bind to one parameter from that overload . :( This seems to be the assumption in a lot of places in IDE code.

**Comment by CyrusNajmabadi** [comment]
I'm not sure the IDE makes many assumptions here.  The general idea is: welp... it's error code, so the candidates are best effort.  we'll just use them to inform some info, with the hope that it's somewhat reasonable in practice across enough scenarios.

**Comment by Neme12** [comment]
@CyrusNajmabadi As you can see in this bug: #5338, some features do. The generated argument comes from a different overload than the return type. In this case, it happens to actually work out anyway but generally this could very easily introduce incorrect code.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi As you can see in this bug: #5338, some features do.

I have no doubt that it has come up.  But my point was that by and large we don't care (and individual case notwithstanding).

> it happens to actually work out anyway but generally this could very easily introduce incorrect code,

The bar for IDE features is far lower than that :)  As per our conversations in the past, the goal is to do a *good* job in the *majority* of common cases.  It's not to do a perfect job in all cases.

**Comment by gafter** [comment]
To answer the original questions...

> Is it an unreasonable assumption that the first candidate symbol would be the one used to gather parameter types?

Yes, that is an unreasonable assumption simply because things don't work that way. We try to find a best parameter type on a parameter-by-parameter basis. We used to do things based on the first candidate, and we found that resulted in really bad error recovery for many situations. See https://github.com/dotnet/roslyn/issues/11619 and the PRS that it references for examples that illustrate why things don't match up so nicely.

> Should they not match?

They don't have anything to do with each other.  We can and do select parameter types from different elements of the list, for example if you pass two lambda expressions, one method has a delegate for the first parameter, and the other method has a delegate for the second parameter.

> Or is there no guaranteed ordering of CandidateSymbols?

Correct, there is no particular order.
